sdks/java/core/src/main/java/org/apache/beam/sdk/options/ValueProvider.java

```diff
@@ -0,0 +1,209 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.options;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.BeanProperty;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.fasterxml.jackson.databind.deser.ContextualDeserializer;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.concurrent.ConcurrentHashMap;
+
+/** {@link ValueProvider} is an interface which abstracts the notion of
+ * fetching a value that may or may not be currently available.  This can be
+ * used to parameterize transforms that only read values in at runtime, for
+ * example.
+ */
+@JsonSerialize(using = ValueProvider.Serializer.class)
+@JsonDeserialize(using = ValueProvider.Deserializer.class)
+public interface ValueProvider<T> {
+  T get();
+
+  /** Whether the contents of this ValueProvider is available to validation
+   * routines that run at graph construction time.
+   */
+  boolean shouldValidate();
+
+  /** {@link StaticValueProvider} is an implementation of ValueProvider that
+   * allows for a static value to be provided.
+   */
+  public static class StaticValueProvider<T> implements ValueProvider<T>, Serializable {
+    private final T value;
+
+    StaticValueProvider(T value) {
+      this.value = value;
+    }
+
+    public static <T> StaticValueProvider<T> of(T value) {
+      StaticValueProvider<T> factory = new StaticValueProvider<>(value);
+      return factory;
+    }
+
+    @Override
+    public T get() {
+      return value;
+    }
+
+    @Override
+    public boolean shouldValidate() {
+      return true;
+    }
+  }
+
+  /** {@link RuntimeValueProvider} is an implementation of ValueProvider that
+   * allows for a value to be provided at execution time rather than at graph
+   * construction time.
+   *
+   * <p>To enforce this contract, if there is no default, users must only call
+   * get() on the worker, which will provide the value of OPTIONS.
```

Comment:
I'm confused about the default thing. I think the text here implies that having a default value makes the option okay to call before job execution. Wouldn't that make it an SVP, which is not borne out by the code above?